Add option to enable test framework to run tests in a specified order.

rhou1 opened the pull request:
The option is enabled with the -rR flag, for repeatRun.
If specified for the first time, it will create a file in framework/resources/queryLog.log
that contains the tests that are run, in the order they are run.
Once this file is created, the next run will run the tests in the order specified
in the file.  The file can also be created manually.
This implementation requires that the tests to be run are also specified by the -s option.
It takes the tests that are normally collected, and re-orders the tests in the order
specified in the queryLog.log file.
This is useful for failures that occur when tests are run in a particular order.
It is mainly useful if the tests are run without concurrency, meaning one at a time.

rhou1 commented:
Chun, can you please review this.

cchang738 commented on framework/src/main/java/org/apache/drill/test/framework/TestDriver.java:
```diff
@@ -188,6 +190,38 @@ public int runTests() throws Exception {
+            reader.close();
+          }
+        }
+        BufferedWriter writer = new BufferedWriter(new FileWriter(new File(
```
This can be optimized, i.e. we probably don't need to write to the file again if it exists already.

cchang738 commented on framework/src/main/java/org/apache/drill/test/framework/TestDriver.java:
```diff
@@ -188,6 +190,38 @@ public int runTests() throws Exception {
+          }
+          writer.close();
+        }
+      }
```
Might be better rearrange the flow a bit as we discussed in person. Thanks.

cchang738 commented on framework/src/main/java/org/apache/drill/test/framework/TestDriver.java:
```diff
@@ -188,6 +190,38 @@ public int runTests() throws Exception {
       LOG.info("> RUNNING TESTS (ITERATION " + i + ")..");
       Collections.shuffle(tests, new Random());
+      if (cmdParam.repeatRun == true) {
+        String queryLogFilename = drillTestDataDir + "/queryLog.log";
```
Can we store the query file inside drill-out dir? This is where we put other tmp files, so keep it consistent would be better.

rhou1 commented:
Hi  Chun,

I have addressed your comments.  Please take a look and let me know if I missed something.

cchang738 commented on framework/src/main/java/org/apache/drill/test/framework/TestDriver.java:
```diff
@@ -187,7 +189,47 @@ public int runTests() throws Exception {
+      File file = new File(queryLogFilename);
+      if ((cmdParam.repeatRun == true) && (file.exists() && (file.length() > 0))) {
+        // run the tests in the requested order
+        try (BufferedReader reader = new BufferedReader(new FileReader(new File(
```
Can we reuse the "file" already created two lines above?

cchang738 commented on framework/src/main/java/org/apache/drill/test/framework/TestDriver.java:
```diff
@@ -187,7 +189,47 @@ public int runTests() throws Exception {
+        Collections.shuffle(tests, new Random());
+        if (cmdParam.repeatRun == true) {
+          // record the tests and the order in which they are run
+          BufferedWriter writer = new BufferedWriter(new FileWriter(new File(
```
Same here, reuse "file".

cchang738 commented on framework/src/main/java/org/apache/drill/test/framework/TestDriver.java:
```diff
@@ -187,7 +189,47 @@ public int runTests() throws Exception {
+        // run the tests in the requested order
+        try (BufferedReader reader = new BufferedReader(new FileReader(new File(
+                                     queryLogFilename)))) {
+          List<DrillTest> tests_copy = Lists.newArrayList(tests);
```
Here can we do the switch at the end after you build the list? No copy needed just reassign?

rhou1 commented:
Hi Chun,

I addressed your comments.  Please take a look and let me know if I missed something.
I also added some code to inform the user when a test is in the queryLog file but not in the source or group.  In which case, the test is skipped.  Only tests that are in the source (-s option) and group (-g option) can be run at this time.

rhou1 commented:
I also fixed a minor issue with Random Failures, where the string "Random Failures" appeared in the Results section even if there are no Random Failures.

Agirish commented:
Hey @rhou1 , can you please detail how this is different from current behavior and how we can guarantee the same execution order? 

To clarify, if we already store collected tests in a specified order (say lexicographic), then shouldn't this be deterministic? 

Also, how do we handle cases with concurrency? A given query can run in 5 seconds or in 10 seconds, depending on load and other characteristics. So even if it reads a query list in your case to determine the order, can we be sure it's respected when multiple queries are run in parallel?

-1 until then.

rhou1 commented:
The current behavior is to shuffle the tests and execute them.  So each execution has a different order.

The last line in the commit says this change mainly helps when there is no concurrency.  It is for debugging.

Agirish commented:
Agreed on the specific usecase. Thanks for clarifying. 

With regard to implementation, curious to understand if we could just not shuffle the tests if this parameter is specified - instead sort them. Then we will not require a file to be written into and read from.

rhou1 commented:
If you don't sort/shuffle the tests, then I will guess you run the tests in the same order each time.

The intent of this change is to allow you to run the tests in the order you want to run them in.  For example, when I was testing External Sort, I found a test passed by itself.  But sometimes it failed when I ran the whole set, depending on how the tests were shuffled.  So I wanted to capture the specific order of tests that caused the one test to fail.

Having the file allows you to specify only the tests you want to run.  Ideally, you can narrow it down to a handful of tests.

Agirish commented:
Cool, thanks again. +1

I think this serves a narrow use case, but if it's helpful then we should get it in.

cchang738 commented on framework/src/main/java/org/apache/drill/test/framework/TestDriver.java:
```diff
@@ -187,7 +189,55 @@ public int runTests() throws Exception {
+            writer.close();
+          }
+        }
+      }
```
Can you see if you can fit your code into the following logic structure:

if (cmdParam.repeatRun) {
  if (queryFile.length() > 0) {
    ...
  } else {
    ...
  }
}
original code...

cchang738 commented on framework/src/main/java/org/apache/drill/test/framework/TestDriver.java:
```diff
@@ -276,14 +326,25 @@ public int runTests() throws Exception {
+      }
       LOG.info(LINE_BREAK);
       LOG.info("Summary");
+      if ((cmdParam.repeatRun == true) && (queryFile.exists() && (queryFile.length() > 0))) {
```
Can queryFile.length change? For example, if you started without the file but created afterwards, will this evaluate to true during the same run?

rhou1 commented:
Hi Chun,

I updated the code based on your comments.  Please take a look.

rhou1 commented:
Added try logic for BufferedWriter.  Chun, please take a look.

rhou1 commented:
Reorganized the code a bit.  Chun, please take a look.

cchang738 commented:
Did you rerun pre-commit test after the final revision? If yes, +1.

rhou1 commented:
I made another change.  I no longer remove the query file from the "test" list at line 210.  This allows a test to be run more than once.

pre-commit test is here:
http://10.10.104.91:8080/job/Framework-Regression/1153/console

cchang738 commented:
+1.
